src/Security/Authentication/Negotiate/src/Internal/LdapAdapter.cs

```diff
@@ -59,30 +61,55 @@ public static async Task RetrieveClaimsAsync(LdapSettings settings, ClaimsIdenti
                 logger.LogWarning($"More than one response received for query: {filter} with distinguished name: {distinguishedName}");
             }
 
-            var userFound = searchResponse.Entries[0]; //Get the object that was found on ldap
+            var userFound = searchResponse.Entries[0]; // Get the object that was found on ldap
             var memberof = userFound.Attributes["memberof"]; // You can access ldap Attributes with Attributes property
 
-            foreach (var group in memberof)
+            // Get the user SID
+            var userSID = userFound.Attributes["objectsid"];
+            if (userSID is not null)
             {
-                // Example distinguished name: CN=TestGroup,DC=KERB,DC=local
-                var groupDN = $"{Encoding.UTF8.GetString((byte[])group)}";
-                var groupCN = DistinguishedNameSeparator().Split(groupDN)[0].Substring("CN=".Length);
-
-                if (!settings.IgnoreNestedGroups)
+                if (userSID.Count == 1)
                 {
-                    GetNestedGroups(settings.LdapConnection, identity, distinguishedName, groupCN, logger, retrievedClaims, new HashSet<string>());
+                    var usid = ParseSID((byte[])userSID[0]);
+                    if (usid is not null)
+                    {
+                        retrievedClaims.Add(new KeyValuePair<string, string>(ClaimTypes.Sid, usid.ToString()));
```

Comment:
Is everyone who's using `EnableLdap` going to want `Sid` and `GroupSid`? This was requested by #31959, but it seems like most people have gotten by fine without it. And it does waste space in the cache.

Also, should this include all of the SID claims mentioned by the issue?

```
primarysid
primarygroupsid
groupsid
denyonlysid
```

There's even more that aren't listed like `denyonlyprimarysid` and `denyonlyprimarygroupsid`.

I think this should be opt-in or at the very least opt-out which will require new public API. You can use our [API proposal issue template](https://github.com/dotnet/aspnetcore/issues/new?assignees=&labels=api-suggestion&projects=&template=30_api_proposal.md&title=) to file a new API proposal issue, or copy the template and fill it out in a new comment on #55705. We can add the `api-ready-for-review` label once it's ready.

Reply:
I think I figured out how to add `primarygroupsid` by constructing it from the RID, but I could not find much useful information about `denyonlysid` (unfortunately my time to look into this is quite restricted so if anyone could point me in the right direction I would be very grateful).